integration/v3_grpc_test.go

```diff
@@ -193,3 +195,225 @@ func TestV3DeleteRange(t *testing.T) {
 		clus.Terminate(t)
 	}
 }
+
+// TestV3WatchFromCurrentRevision tests Watch APIs from current revision.
+func TestV3WatchFromCurrentRevision(t *testing.T) {
+	tests := []struct {
+		initialKeys   []string
+		initialValue  string
+		watchRequests []pb.WatchRequest
+		events        []storagepb.Event
+	}{
+		// watch the key
+		{
+			[]string{"foo"},
+			"bar",
+			[]pb.WatchRequest{{CreateRequest: &pb.WatchCreateRequest{Key: []byte("foo")}}},
+			[]storagepb.Event{
+				{
+					Kv: &storagepb.KeyValue{Key: []byte("foo"), Value: []byte("bar"), CreateRevision: 1, ModRevision: 1, Version: 1},
+				},
+			},
+		},
+		// watch the prefix, matching
+		{
+			[]string{"foolong"},
+			"bar",
+			[]pb.WatchRequest{{CreateRequest: &pb.WatchCreateRequest{Prefix: []byte("foo")}}},
+			[]storagepb.Event{
+				{
+					Kv: &storagepb.KeyValue{Key: []byte("foolong"), Value: []byte("bar"), CreateRevision: 1, ModRevision: 1, Version: 1},
+				},
+			},
+		},
+		// watch the prefix, non-matching
+		{
+			[]string{"foo"},
+			"bar",
+			[]pb.WatchRequest{{CreateRequest: &pb.WatchCreateRequest{Prefix: []byte("foobar")}}},
+			[]storagepb.Event{},
+		},
+	}
+
+	for i, tt := range tests {
+		clus := newClusterGRPC(t, &clusterConfig{size: 3})
+
+		wAPI := pb.NewWatchClient(clus.RandConn())
+		wStream, err := wAPI.Watch(context.TODO())
+		if err != nil {
+			t.Errorf("#%d: wAPI.Watch error: %v", i, err)
+		}
+
+		for _, r := range tt.watchRequests {
+			if err := wStream.Send(&r); err != nil {
+				t.Errorf("#%d: wStream.Send error: %v", i, err)
+			}
+		}
+
+		kvc := pb.NewKVClient(clus.RandConn())
+		for _, k := range tt.initialKeys {
+			reqput := &pb.PutRequest{Key: []byte(k), Value: []byte(tt.initialValue)}
+			if _, err := kvc.Put(context.TODO(), reqput); err != nil {
+				t.Errorf("#%d: couldn't put key (%v)", i, err)
+			}
+		}
+
+		copied := []storagepb.Event{}
+		cnt := 0
+		for cnt != len(tt.events) {
+			resp, err := wStream.Recv()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				t.Errorf("#%d: wStream.Recv error: %v", i, err)
+			}
+
+			switch {
+			case resp.Created:
+				t.Logf("#%d: watcher created: id %08x\n", i, resp.WatchId)
+			default:
+				if len(resp.Events) != len(tt.events) {
+					t.Errorf("#%d: len(resp.Events) got = %d, want = %d", i, len(resp.Events), len(tt.events))
+				}
+
+				for _, r := range resp.Events {
+					if r != nil && r.Kv != nil {
+						copied = append(copied, *r)
+						cnt++
+					}
+				}
+			}
+		}
+
+		if !reflect.DeepEqual(copied, tt.events) {
+			t.Errorf("#%d: got = %+v, want = %+v", i, copied, tt.events)
+		}
+
+		for _, ev := range copied {
+			t.Logf("%s: %s %s\n", ev.Type, string(ev.Kv.Key), string(ev.Kv.Value))
+		}
+
+		// can't defer because tcp ports will be in use
+		clus.Terminate(t)
+	}
+}
+
+// TestV3WatchUnsynced tests Watch APIs when watchers are unsynced.
+func TestV3WatchUnsynced(t *testing.T) {
```

Comment:
test failed on this one. Let's move this one to another pr and focus on merging the sync one first.